### PR TITLE
Stub generator

### PR DESCRIPTION
**Pull request by sstanculeanu**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- script to generate stubs from interfaces
  
## Proposed Changes
- 
- 
- 

## Testing procedure
- 
- 
-

### CODE REVIEW

File: scripts/generators/stubGenerator.sh [hunk 1]
```diff
@@ -0,0 +1,333 @@
+# This script generates a stub from a given interface
```
Comment (valentin-lup) [inline scripts/generators/stubGenerator.sh:1]: What works or not depends on the shell used. Add a shebang. (#!/bin/bash or #!/bin/sh)

File: scripts/generators/stubGenerator.sh [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+  lenOfStubName=${#stubName}
+  stubFileName=$firstChar${stubName:1:$lenOfStubName}
+
+  stubPath="$stubDir/$stubFileName.go"
+  rm -rf "$stubPath"
+
+  isInterfaceMethod=false
+  declare -a methodsArr
+
+  readInterfaceFile
```
Comment (valentin-lup) [inline scripts/generators/stubGenerator.sh:35]: The function definition must be placed before any calls to the function !!! (bash is not a compiler)

File: scripts/generators/stubGenerator.sh [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+  createStubStructure
```
Comment (valentin-lup) [inline scripts/generators/stubGenerator.sh:36]: The function definition must be placed before any calls to the function !!! (bash is not a compiler)

File: scripts/generators/stubGenerator.sh [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+  createStubMethods
```
Comment (valentin-lup) [inline scripts/generators/stubGenerator.sh:37]: The function definition must be placed before any calls to the function !!! (bash is not a compiler)

File: scripts/generators/stubGenerator.sh [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+generateStub() {
+  interfaceName=$1
+  filePath=$2"/interface.go"
+  stubDir=$3
+
+  [ ! -d "$2" ] && echo "Source directory for interface DOES NOT exists." && exit
+  [ ! -f "$filePath" ] && echo "Source interface.go file DOES NOT exists." && exit
+  [ ! -d "$stubDir" ] && echo "Destination directory DOES NOT exists." && exit
+
+  extractPackageName
```
Comment (valentin-lup) [inline scripts/generators/stubGenerator.sh:18]: The function definition must be placed before any calls to the function !!! (bash is not a compiler)

File: scripts/generators/stubGenerator.sh [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+  if [[ ${#updatedParameters} != 1 ]]; then
+    # remove the last ", " added with last param
+    updatedParameters=${updatedParameters::-2}
+  fi
+
+  updatedParameters+=")"
+}
+
+removeCommentsFromMethodLine() {
```
Comment (valentin-lup) [inline scripts/generators/stubGenerator.sh:318]: The function definition must be placed before any calls to the function !!! (bash is not a compiler)

File: scripts/generators/stubGenerator.sh [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+  if [[ "$method" == *"//"* ]]; then
+    method=${method%%"//"*}
+  fi
+}
+
+main() {
```
Comment (valentin-lup) [inline scripts/generators/stubGenerator.sh:324]: Bash does not need/use a main() function.
Not an incorrect use of a main function, but not really needed...

**Review by valentin-lup — CHANGES_REQUESTED** [review]

File: scripts/generators/stubGenerator.sh (hunk 7 above)
Comment (sstanculeanu) [reply scripts/generators/stubGenerator.sh]: removed the method

**Review by valentin-lup — APPROVED** [review]

**Review by iulianpascalau — COMMENTED** [review]
I think we need another condition when the method needs to output several params.
Please test with the following:
```
./stubGenerator.sh TransactionCoordinator ../../process ../../process
```
as it outputs this
```
../../process/transactionCoordinatorStub.go:45:86: expected declaration, found ')'
exit status 2
```

File: scripts/generators/stubGenerator.sh [hunk 8]
```diff
@@ -254,24 +302,52 @@ createStubMethods() {
...
+      #   all parameters until the closing brace of the func ptr parameters
+      rawParams=${method%")"*}
+      rawParams=${rawParams#*"("}
+
+      declare -a paramNames=()
+      declare -a paramTypes=()
+
+      extractParametersAndTypes "$rawParams"
+
+      #for i in "${!paramNames[@]}"
```
Comment (iulianpascalau) [inline scripts/generators/stubGenerator.sh:320]: commented code

Comment (sstanculeanu) [reply scripts/generators/stubGenerator.sh]: removed

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by valentin-lup — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]
tested - works :+1:

**Pull request merged**